src/dataplane-protocol/src/batch.rs

```diff
@@ -36,10 +33,7 @@ pub const BATCH_PREAMBLE_SIZE: usize = size_of::<Offset>()     // Offset
 pub const BATCH_FILE_HEADER_SIZE: usize = BATCH_PREAMBLE_SIZE + BATCH_HEADER_SIZE;
 
 #[derive(Default, Debug)]
-pub struct Batch<R>
-where
-    R: BatchRecords,
-{
+pub struct Batch<R = Vec<Record>> {
```

Comment:
Sure. So we had `type DefaultBatch = Batch<Vec<Record>>` prior to this. Everywhere in the codebase that we were using "Batch", we were actually using `DefaultBatch`. When using `struct Batch<R = Vec<Record>>`, all it says is that if somebody uses a type annotation like `fn get_batch(batch: Batch)`, then it is automatically interpreted to mean `fn get_batch(batch: Batch<Vec<Record>>)`. However, it does not force you to _always_ use that default generic parameter. If you want, you could manually specify the generic, such as `fn get_batch(batch: Batch<HashSet<Record>>)` (for example).

So this does not "force" the struct to use this generic, rather it just allows you to optionally omit the generic at the type-site and it will be defaulted to `Vec<Record>`.

An interesting example of this pattern of usage is with the Vec type itself: https://doc.rust-lang.org/stable/src/alloc/vec/mod.rs.html#374-377

It was recently given a new type signature, `struct Vec<T, A: Allocator = Global>`. This new type parameter does not actually affect the day-to-day usage of the type (because you are not required to manually provide the extra generic), but it optionally allows you to specify a custom allocator, e.g. if you wanted to use Vec in an embedded device.

Reply:
Sounds good.  Didn't realize you could provide default type.  This make it reading code easier.  Good find